internal/controller/supervisorconfig/upstreamwatcher/ldap_upstream_watcher.go

```diff
@@ -0,0 +1,367 @@
+		finishedConfigCondition := c.validateFinishedConfig(ctx, upstream, config, currentSecretVersion)
+		// nil when there is no need to update this condition.
+		if finishedConfigCondition != nil {
+			conditions = append(conditions, finishedConfigCondition)
+		}
+	}
+
+	hadErrorCondition := c.updateStatus(ctx, upstream, conditions)
+	if hadErrorCondition {
+		return nil
+	}
+
+	return upstreamldap.New(*config)
+}
+
+func (c *ldapWatcherController) validateTLSConfig(upstream *v1alpha1.LDAPIdentityProvider, config *upstreamldap.ProviderConfig) *v1alpha1.Condition {
+	tlsSpec := upstream.Spec.TLS
+	if tlsSpec == nil {
+		return c.validTLSCondition(noTLSConfigurationMessage)
+	}
+	if len(tlsSpec.CertificateAuthorityData) == 0 {
+		return c.validTLSCondition(loadedTLSConfigurationMessage)
+	}
+
+	bundle, err := base64.StdEncoding.DecodeString(tlsSpec.CertificateAuthorityData)
+	if err != nil {
+		return c.invalidTLSCondition(fmt.Sprintf("certificateAuthorityData is invalid: %s", err.Error()))
+	}
+
+	ca := x509.NewCertPool()
+	ok := ca.AppendCertsFromPEM(bundle)
+	if !ok {
+		return c.invalidTLSCondition(fmt.Sprintf("certificateAuthorityData is invalid: %s", errNoCertificates))
+	}
+
+	config.CABundle = bundle
+	return c.validTLSCondition(loadedTLSConfigurationMessage)
+}
+
+func (c *ldapWatcherController) validateFinishedConfig(ctx context.Context, upstream *v1alpha1.LDAPIdentityProvider, config *upstreamldap.ProviderConfig, currentSecretVersion string) *v1alpha1.Condition {
+	ldapProvider := upstreamldap.New(*config)
+
+	if hasPreviousSuccessfulConditionForCurrentSpecGenerationAndSecretVersion(upstream, currentSecretVersion) {
+		return nil
+	}
+
+	testConnectionTimeout, cancelFunc := context.WithTimeout(ctx, testLDAPConnectionTimeout)
+	defer cancelFunc()
+
+	if len(upstream.Spec.DryRunAuthenticationUsername) > 0 {
+		return c.dryRunAuthentication(testConnectionTimeout, upstream, ldapProvider, currentSecretVersion)
+	}
+
+	return c.testConnection(testConnectionTimeout, upstream, config, ldapProvider, currentSecretVersion)
+}
+
+func (c *ldapWatcherController) testConnection(
+	ctx context.Context,
+	upstream *v1alpha1.LDAPIdentityProvider,
+	config *upstreamldap.ProviderConfig,
+	ldapProvider *upstreamldap.Provider,
+	currentSecretVersion string,
+) *v1alpha1.Condition {
+	err := ldapProvider.TestConnection(ctx)
+	if err != nil {
+		return &v1alpha1.Condition{
+			Type:   typeLDAPConnectionValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonLDAPConnectionError,
+			Message: fmt.Sprintf(`could not successfully connect to "%s" and bind as user "%s": %s`,
+				config.Host, config.BindUsername, err.Error()),
+		}
+	}
+
+	return &v1alpha1.Condition{
+		Type:   typeLDAPConnectionValid,
+		Status: v1alpha1.ConditionTrue,
+		Reason: reasonSuccess,
+		Message: fmt.Sprintf(`successfully able to connect to "%s" and bind as user "%s" [validated with Secret "%s" at version "%s"]`,
```

Comment:
Can we just use a different condition to store this?

The more correct way to do this is to add something like this to the `Spec` of the LDAP IDP (though I do not think we should do that at this time):

```go
// observedConfig holds a sparse config that the controller has observed from the cluster state.
// It exists in spec because it is an input to the level for the operator
// +optional
// +nullable
// +kubebuilder:pruning:PreserveUnknownFields
ObservedConfig runtime.RawExtension `json:"observedConfig"`
```

Reply:
Actually thinking on this more, this is not a safe construct.  We are trying to store the resource version from an informer cache to an API object.  The informer caches across our replicas are not guaranteed to be in sync, meaning the controllers could easily start fighting over this field if the informer ever wedges.

At a minimum we need to do a live get against the cluster.

I think a better approach would be some way of storing the resource version in memory as part of the cache entry.  This if it changes, the cache should become invalid.

Reply:
Hmm, really? That's subtle.

The docs for `ResourceVersion` say:
```
	// An opaque value that represents the internal version of this object that can
	// be used by clients to determine when objects have changed. May be used for optimistic
	// concurrency, change detection, and the watch operation on a resource or set of resources.
	// Clients must treat these values as opaque and passed unmodified back to the server.
	// They may only be valid for a particular resource or set of resources.
```

That makes it sound like it is a value assigned by the server, so it would not come from the informer. Different informers on different pods might temporarily see different version of the secret based on how recent their cache was updated, but that should be okay because they will eventually catch up with each other and agree on what is the latest version, and the controller syncs on any changes to the secret, so it will eventually test and record the result from the correct combination of spec version with secret version.

Reply:
I see what you mean about the pods fighting over this condition field if one of them has a broken informer.

I was trying to avoid the even worse edge case that could happen if the record of which secret was validated were stored in RAM... if the pod restarts then it loses that record, and during the next sync would try to revalidate the user's spec. However, if the LDAP server/network has a transient error during that time period, then the controller's best action from its point of view would be to pull the IDP out of the authenticator cache because it thinks that the spec in invalid. However, there would actually be nothing wrong with the spec and there is no reason to stop accepting login attempts for that IDP.

Reply:
> However, if the LDAP server/network has a transient error during that time period, then the controller's best action from its point of view would be to pull the IDP out of the authenticator cache because it thinks that the spec in invalid.

The controller should leave the LDAP IDP in the cache as long as it can actually construct the object.

Reply:
I believe the simplest solution for now is to have the in-memory cache track the RV and to never write the RV to the status.  We should keep the LDAP IDP in the cache if at all possible as I do not see any harm in letting someone try to use it.

Reply:
Note that RV can change even if the secret data is unchanged (storage migration, custom controllers, etc).  Thus our code must tolerate that.

Reply:
Changed in f5bf8978a364860c691d1d5b2d43491aef059ec1 to use an in-memory cache of the Secret's ResourceVersion.

- When the dial/bind probe fails, write the Condition to the status but go ahead and still allow end-user login attempts using that LDAP provider, effectively making the Condition just a warning.
- When the Secret's RV changes, go ahead and attempt the dial/bind probe again, which is sometimes unnecessary because the actual username/password inside the Secret may not have changed, but it doesn't hurt anything to try again.
- When the pod restarts, attempt the dial/bind probe again, which again doesn't hurt anything.

cc @mattmoyer